src-extras/Database/LSMTree/Extras/NoThunks.hs

```diff
@@ -315,9 +315,33 @@ deriving anyclass instance NoThunks a => NoThunks (SMaybe a)
   PageAcc
 -------------------------------------------------------------------------------}
 
-deriving stock instance Generic (PageAcc s)
-deriving anyclass instance Typeable s
-                        => NoThunks (PageAcc s)
+instance Typeable s => NoThunks (PageAcc s) where
+  showTypeOf (p :: Proxy (PageAcc s)) = show $ typeRep p
+  wNoThunks ctx
+    pa@(PageAcc (a :: MutablePrimArray s Int) (b :: MutablePrimArray s Word64)
+                (c :: MutablePrimArray s Word64) (d :: MutablePrimArray s Word64)
+                (e :: MutablePrimArray s Word32) (f :: VM.MVector s SerialisedKey)
+                (g :: VM.MVector s SerialisedValue))
+    = allNoThunks [
+          noThunks ctx a
+        , noThunks ctx b
+        , noThunks ctx c
+        , noThunks ctx d
+        , noThunks ctx e
+          -- the keys and values vectors array may contain uninitialised
+          -- entries, so we should check only up to the number of keys/values
+          -- that are actually stored in the PageAcc.
```

Comment:
And then maybe we should check that the values outside of the currently-used range are some specific thunk to represent uninitialised. Afterall, no thunks tests are about avoiding memory leaks, and keeping a reference in an "unused" part of the array is a good source for a memory leak.